src/ahbicht/validation/validation.py

```diff
@@ -0,0 +1,357 @@
+            segment_group, parent_segment_group_requirement, soll_is_required
+        )
+
+    validation_results_in_context = [
+        ValidationResultInContext(discriminator=segment_group.discriminator, validation_result=segment_group_validation)
+    ]
+
+    tasks = []
+
+    # validation of child_segment_group s
+    if segment_group.segment_groups:
+        for child_segment_group in segment_group.segment_groups:
+            tasks.append(
+                validate_segment_group(
+                    child_segment_group,
+                    segment_group_validation.requirement_validation,
+                    soll_is_required,
+                )
+            )
+
+    # validation of child segments
+    if segment_group.segments:
+        for segment in segment_group.segments:
+            tasks.append(
+                validate_segment(
+                    segment,
+                    segment_group_validation.requirement_validation,
+                    soll_is_required,
+                )
+            )
+
+    validation_results_of_children = await asyncio.gather(*tasks)
+
+    for sublist in validation_results_of_children:
+        for item in sublist:
+            validation_results_in_context.append(item)
+
+    return validation_results_in_context
+
+
+async def validate_segment(
+    segment: Segment,
+    segment_group_requirement: Optional[RequirementValidationValue] = None,
+    soll_is_required: bool = True,
+) -> List[ValidationResultInContext]:
+    """
+    Validates a segment and the data elements it contains.
+    :param segment: the segment that should be validated
+    :param segment_group_requirement: the requirement of the segment's segment_group, e.g. IS_REQUIRED
+    :param soll_is_required: true (default) if SOLL should be handled like MUSS, false if it should be handled like KANN
+    :return: List of ValidationResultInContext of the segment and the data elements it contains
+    """
+
+    # validation of this segment
+    if segment_group_requirement is RequirementValidationValue.IS_FORBIDDEN:
+        segment_validation = SegmentLevelValidationResult(
+            requirement_validation=RequirementValidationValue.IS_FORBIDDEN
+        )
+    else:
+        segment_validation = await get_segment_level_requirement_validation_value(
+            segment, segment_group_requirement, soll_is_required
+        )
+
+    validation_results_in_context = [
+        ValidationResultInContext(discriminator=segment.discriminator, validation_result=segment_validation)
+    ]
+
+    tasks = []
+
+    # validation of this segments dataelements
+    for dataelement in segment.data_elements:
+        tasks.append(validate_dataelement(dataelement, segment_validation.requirement_validation))
+
+    validation_results_in_context_dataelements = await asyncio.gather(*tasks)
+
+    return [*validation_results_in_context, *validation_results_in_context_dataelements]
```

Comment:
das verstehe ich noch nicht ganz: am ende sind das ergebnis des segments und der untergeordneten datenelement in einer flachen liste. passt das so fürs FE? wir verlieren da ja die information was worunter hängt.

Reply:
Das Frontend erwartet eine flache Liste, by design ;)

Reply:
Maus Strukur macht im Frontend (BO4E) keinen Sinn mehr.